
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Food DNA Metabarcoding analysis">
      
      
        <meta name="author" content="Grégoire Denay and Marc Höppner">
      
      
        <link rel="canonical" href="https://bio-raum.github.io/FooDMe2/dev/user_doc/usage/">
      
      
        <link rel="prev" href="../installation/">
      
      
        <link rel="next" href="../pipeline/">
      
      
      <link rel="icon" href="../../img/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.11">
    
    
      
        <title>Usage - FooDMe2 Documentation</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.4af4bdda.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      

  
  
  
  
  <style>:root{--md-annotation-icon:url('data:image/svg+xml;charset=utf-8,%3Csvg%20xmlns%3D%22http%3A//www.w3.org/2000/svg%22%20viewBox%3D%220%200%2024%2024%22%3E%3Cpath%20d%3D%22m15.07%2011.25-.9.92C13.45%2012.89%2013%2013.5%2013%2015h-2v-.5c0-1.11.45-2.11%201.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41a2%202%200%200%200-2-2%202%202%200%200%200-2%202H8a4%204%200%200%201%204-4%204%204%200%200%201%204%204%203.2%203.2%200%200%201-.93%202.25M13%2019h-2v-2h2M12%202A10%2010%200%200%200%202%2012a10%2010%200%200%200%2010%2010%2010%2010%200%200%200%2010-10c0-5.53-4.5-10-10-10%22/%3E%3C/svg%3E');}</style>


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="deep-orange">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#usage-information" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
          <aside class="md-banner md-banner--warning">
            <div class="md-banner__inner md-grid md-typeset">
              
  You're not viewing the latest version.
  <a href="../../.."> 
    <strong>Click here to go to latest.</strong>
  </a>

            </div>
            <script>var el=document.querySelector("[data-md-component=outdated]"),base=new URL("../.."),outdated=__md_get("__outdated",sessionStorage,base);!0===outdated&&el&&(el.hidden=!1)</script>
          </aside>
        
      </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="FooDMe2 Documentation" class="md-header__button md-logo" aria-label="FooDMe2 Documentation" data-md-component="logo">
      
  <img src="../../img/pipelinelogo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            FooDMe2 Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Usage
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="deep-orange"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0 8a5 5 0 0 1-5-5 5 5 0 0 1 5-5 5 5 0 0 1 5 5 5 5 0 0 1-5 5m0-12.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="deep-orange"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 9a3 3 0 0 1 3 3 3 3 0 0 1-3 3 3 3 0 0 1-3-3 3 3 0 0 1 3-3m0-4.5c5 0 9.27 3.11 11 7.5-1.73 4.39-6 7.5-11 7.5S2.73 16.39 1 12c1.73-4.39 6-7.5 11-7.5M3.18 12a9.821 9.821 0 0 0 17.64 0 9.821 9.821 0 0 0-17.64 0"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/bio-raum/FooDMe2" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../.." class="md-tabs__link">
          
  
  
  Home

        </a>
      </li>
    
  

      
        
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../quickstart/" class="md-tabs__link">
          
  
  
  User Guide

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../methods/methods/" class="md-tabs__link">
          
  
  
  Methods

        </a>
      </li>
    
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../help/migration/" class="md-tabs__link">
          
  
  
  Help

        </a>
      </li>
    
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../developer/" class="md-tabs__link">
        
  
  
    
  
  Developer Guide

      </a>
    </li>
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../about/software/" class="md-tabs__link">
          
  
  
  About

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="FooDMe2 Documentation" class="md-nav__button md-logo" aria-label="FooDMe2 Documentation" data-md-component="logo">
      
  <img src="../../img/pipelinelogo.png" alt="logo">

    </a>
    FooDMe2 Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/bio-raum/FooDMe2" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_1" >
        
          
          <label class="md-nav__link" for="__nav_1" id="__nav_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            Home
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Welcome
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../about/contributing/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Contibuting
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../about/changelog/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Release notes
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    User Guide
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            User Guide
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../quickstart/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Quickstart
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../requirements/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Requirements
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Usage
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Usage
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#running-the-pipeline" class="md-nav__link">
    <span class="md-ellipsis">
      Running the pipeline
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#removing-temporary-data" class="md-nav__link">
    <span class="md-ellipsis">
      Removing temporary data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#specifying-a-pipeline-version" class="md-nav__link">
    <span class="md-ellipsis">
      Specifying a pipeline version
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#updating-to-a-new-release" class="md-nav__link">
    <span class="md-ellipsis">
      Updating to a new release
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#running-a-test" class="md-nav__link">
    <span class="md-ellipsis">
      Running a test
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#command-line-option" class="md-nav__link">
    <span class="md-ellipsis">
      Command-line option
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Command-line option">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#basic-options" class="md-nav__link">
    <span class="md-ellipsis">
      Basic options
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pcr-primers" class="md-nav__link">
    <span class="md-ellipsis">
      PCR primers
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#database" class="md-nav__link">
    <span class="md-ellipsis">
      Database
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#advanced-options" class="md-nav__link">
    <span class="md-ellipsis">
      Advanced options
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Advanced options">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#benchmarking" class="md-nav__link">
    <span class="md-ellipsis">
      Benchmarking
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pcr-primer-trimming" class="md-nav__link">
    <span class="md-ellipsis">
      PCR primer trimming
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#read-trimming" class="md-nav__link">
    <span class="md-ellipsis">
      Read trimming
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#clustering-configuration" class="md-nav__link">
    <span class="md-ellipsis">
      Clustering configuration
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#blast-configuration" class="md-nav__link">
    <span class="md-ellipsis">
      BLAST configuration
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../pipeline/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Pipeline
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../output/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Output
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Methods
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Methods
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/methods/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Standard methods
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/amniotes_dobrovolny/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Illumina Mammals and birds
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Help
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Help
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../help/migration/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Migration
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../help/troubleshooting/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Troobleshooting
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../developer/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Developer Guide
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    About
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            About
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../about/software/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Software
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../about/license/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    License
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#running-the-pipeline" class="md-nav__link">
    <span class="md-ellipsis">
      Running the pipeline
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#removing-temporary-data" class="md-nav__link">
    <span class="md-ellipsis">
      Removing temporary data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#specifying-a-pipeline-version" class="md-nav__link">
    <span class="md-ellipsis">
      Specifying a pipeline version
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#updating-to-a-new-release" class="md-nav__link">
    <span class="md-ellipsis">
      Updating to a new release
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#running-a-test" class="md-nav__link">
    <span class="md-ellipsis">
      Running a test
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#command-line-option" class="md-nav__link">
    <span class="md-ellipsis">
      Command-line option
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Command-line option">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#basic-options" class="md-nav__link">
    <span class="md-ellipsis">
      Basic options
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pcr-primers" class="md-nav__link">
    <span class="md-ellipsis">
      PCR primers
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#database" class="md-nav__link">
    <span class="md-ellipsis">
      Database
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#advanced-options" class="md-nav__link">
    <span class="md-ellipsis">
      Advanced options
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Advanced options">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#benchmarking" class="md-nav__link">
    <span class="md-ellipsis">
      Benchmarking
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pcr-primer-trimming" class="md-nav__link">
    <span class="md-ellipsis">
      PCR primer trimming
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#read-trimming" class="md-nav__link">
    <span class="md-ellipsis">
      Read trimming
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#clustering-configuration" class="md-nav__link">
    <span class="md-ellipsis">
      Clustering configuration
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#blast-configuration" class="md-nav__link">
    <span class="md-ellipsis">
      BLAST configuration
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="usage-information">Usage information</h1>
<h2 id="running-the-pipeline">Running the pipeline</h2>
<p>Please see our <a href="../installation/">installation guide</a> to learn how to set up this pipeline first.</p>
<p>A basic execution of the pipeline looks as follows:</p>
<div class="tabbed-set tabbed-alternate" data-tabs="1:2"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><input id="__tabbed_1_2" name="__tabbed_1" type="radio" /><div class="tabbed-labels"><label for="__tabbed_1_1">Built-in profile</label><label for="__tabbed_1_2">Site-specific profile</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="language-bash highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>bio-raum/FooDMe2<span class="w"> </span><span class="se">\</span>
</span><span id="__span-0-2"><a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a><span class="w">  </span>-profile<span class="w"> </span>apptainer<span class="w"> </span><span class="se">\ </span><span class="c1"># (1)!</span>
</span><span id="__span-0-3"><a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a><span class="w">  </span>-r<span class="w"> </span>main<span class="w"> </span><span class="se">\ </span><span class="c1"># (2)!</span>
</span><span id="__span-0-4"><a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a><span class="w">  </span>--input<span class="w"> </span>samples.tsv<span class="w"> </span><span class="se">\</span>
</span><span id="__span-0-5"><a id="__codelineno-0-5" name="__codelineno-0-5" href="#__codelineno-0-5"></a><span class="w">  </span>--reference_base<span class="w"> </span>/path/to/references<span class="w"> </span><span class="se">\ </span><span class="c1"># (3)!</span>
</span><span id="__span-0-6"><a id="__codelineno-0-6" name="__codelineno-0-6" href="#__codelineno-0-6"></a><span class="w">  </span>--run_name<span class="w"> </span>pipeline-test<span class="w"> </span><span class="se">\</span>
</span><span id="__span-0-7"><a id="__codelineno-0-7" name="__codelineno-0-7" href="#__codelineno-0-7"></a><span class="w">  </span>--primer_set<span class="w"> </span>amniotes_dobrovolny
</span></code></pre></div>
<ol>
<li>In this example, the pipeline will assume it runs on a single computer with the conda engine. Available options to provision software are documented in the <a href="../installation/">installation section</a>.</li>
<li>We highly recommend pinning a release number(e.g. <code>-r 1.0.0</code>) instead of using the latest commit.</li>
<li><code>path_to_references</code> corresponds to the location in which you have <a href="../installation/">installed</a> the pipeline references.</li>
</ol>
</div>
<div class="tabbed-block">
<div class="language-bash highlight"><pre><span></span><code><span id="__span-1-1"><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>bio-raum/FooDMe2
</span><span id="__span-1-2"><a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a><span class="w">  </span>-profile<span class="w"> </span>myprofile<span class="w"> </span><span class="se">\ </span><span class="c1"># (1)!</span>
</span><span id="__span-1-3"><a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a><span class="w">  </span>-r<span class="w"> </span>main<span class="w"> </span><span class="se">\ </span><span class="c1"># (2)!</span>
</span><span id="__span-1-4"><a id="__codelineno-1-4" name="__codelineno-1-4" href="#__codelineno-1-4"></a><span class="w">  </span>--input<span class="w"> </span>samples.tsv<span class="w"> </span><span class="se">\</span>
</span><span id="__span-1-5"><a id="__codelineno-1-5" name="__codelineno-1-5" href="#__codelineno-1-5"></a><span class="w">  </span>--run_name<span class="w"> </span>pipeline-test<span class="w"> </span><span class="se">\</span>
</span><span id="__span-1-6"><a id="__codelineno-1-6" name="__codelineno-1-6" href="#__codelineno-1-6"></a><span class="w">  </span>--primer_set<span class="w"> </span>amniotes_dobrovolny
</span></code></pre></div>
<ol>
<li>In this example, both <code>--reference_base</code> and the choice of software provisioning are already set in the  configuration <code>lsh</code> and don't have to provided as command line argument. In addition, in your site-specific configuration, you can set additional site-specific parameters, such as your local resource manager, node configuration (CPU, RAM, wall time), desired cache directory for the configured package/container software etc. It is highly recommended to <a href="../installation/">set up</a> such a config file.</li>
<li>We highly recommend pinning a release number(e.g. <code>-r 1.0.0</code>) instead of using the latest commit.</li>
</ol>
</div>
</div>
</div>
<h2 id="removing-temporary-data">Removing temporary data</h2>
<p>Nextflow stores all the process data in a folder structure inside the <code>work</code> directory. All the relevant results are subsequently copied to the designated results folder (<code>--outdir</code>). The work directory is needed to resume completed or failed pipeline runs, but should be removed once you are satisified with the analysis to save space. To do so, run:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-2-1"><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>nextflow<span class="w"> </span>clean<span class="w"> </span>-f
</span></code></pre></div>
<h2 id="specifying-a-pipeline-version">Specifying a pipeline version</h2>
<p>If you are running this pipeline in a production setting, you will want to lock the pipeline to a specific version. This is natively supported through nextflow with the <code>-r</code> argument:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-3-1"><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>bio-raum/FooDMe2<span class="w"> </span>-profile<span class="w"> </span>myprofile<span class="w"> </span>-r<span class="w"> </span><span class="m">1</span>.0.0<span class="w"> </span>&lt;other<span class="w"> </span>options<span class="w"> </span>here&gt;
</span></code></pre></div>
<p>The <code>-r</code> option specifies a github <a href="https://github.com/bio-raum/FooDMe2/releases">release tag</a> or branch, so could also point to <code>main</code> for the very latest code release. Please note that every major release of this pipeline (1.0, 2.0 etc) comes with a new reference data set, which has the be <a href="../installation/">installed</a> separately.</p>
<p>If you are feeling adventurous, you can also provide a specific commit hash, for example: <code>-r 59a710b</code>.</p>
<h2 id="updating-to-a-new-release">Updating to a new release</h2>
<p>To use a new release, you simply have to tell Nextflow to pull the last version from the source (e.g. Github):</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-4-1"><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>nextflow<span class="w"> </span>pull<span class="w"> </span>bio-raum/FooDMe2
</span></code></pre></div>
<p>Then use the <code>-r</code> argument as explained above to run the workflow using the new release.</p>
<h2 id="running-a-test">Running a test</h2>
<p>This pipeline has a built-in test to quickly check that your local setup is working correctly. To run it, do:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-5-1"><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>bio-raum/FooDMe2<span class="w"> </span>-profile<span class="w"> </span>myprofile,test
</span></code></pre></div>
<p>where <code>myprofile</code> can either be a site-specific config file or one of the built-in <a href="../installation/#software-provisioning">profiles</a>. This test requires an active internet connection to download the test data. </p>
<h2 id="command-line-option">Command-line option</h2>
<h3 id="basic-options">Basic options</h3>
<dl>
<dt><code>--input samples.tsv</code> [default = null]</dt>
<dd>
<p>This pipeline expects a TSV-formatted sample sheet (tabulation-delimited text file) to properly pull various meta data through the processes. The required format looks as follows:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-6-1"><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a>sample  fq1 fq2
</span><span id="__span-6-2"><a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a>S100    /path/to/S100_R1.fastq.gz   /path/to/S100_R2.fastq.gz
</span></code></pre></div>
<p>If the pipeline sees more than one set of reads for a given sample ID (i.e. from multi-lane sequencing runs), it will concatenate them automatically at the appropriate time.</p>
<p>This option is the preferred way to provide data to FooDMe2 and it is mutually exclusive with <code>--reads</code>.</p>
<div class="admonition tip automated sample sheet generation">
<p class="admonition-title">Tip</p>
<p>If you want to automatically generate sample sheets from files in a folder, you can use a companion pipeline of FooDMe2:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-7-1"><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>bio-raum/samplesheet<span class="w"> </span>-profile<span class="w"> </span>myprofile<span class="w"> </span>--input<span class="w"> </span>/path/to/folder
</span></code></pre></div>
<p>The folder should contain reads with the extension .fastq.gz or .fq.gz. The pipeline will try to guess the proper grouping rules, but please check the resulting file (results/samples.tsv) for correctness. </p>
</div>
</dd>
<dt><code>--reads</code> [ default = null ]</dt>
<dd>
<p>This option is an alternative (but discouraged!) way to load data into FooDMe2 and expects a wildcard pattern to specify the location of files and how to group them. </p>
<p>Given a set of paired-end reads:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-8-1"><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a>Libary1-S01_R1_001.fastq.gz
</span><span id="__span-8-2"><a id="__codelineno-8-2" name="__codelineno-8-2" href="#__codelineno-8-2"></a>Libary1-S01_R2_001.fastq.gz
</span><span id="__span-8-3"><a id="__codelineno-8-3" name="__codelineno-8-3" href="#__codelineno-8-3"></a>Libary2-S02_R1_001.fastq.gz
</span><span id="__span-8-4"><a id="__codelineno-8-4" name="__codelineno-8-4" href="#__codelineno-8-4"></a>Libary2-S02_R1_001.fastq.gz
</span></code></pre></div>
<p>data can be loaded like so (note the single-quotes around the search pattern!):</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-9-1"><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a>nextflow run bio-raum/FooDMe2 -profile singularity --reads &#39;/path/to/reads/*_R{1,2}_001.fastq.gz&#39;
</span></code></pre></div>
<p>which will be interpreted as two samples, Library1-S01 and Library2-S02, in paired-end configuration. It avoids having to create a samplesheet, but requires a well-constructed wildcard pattern to correctly match all the data as well as provides essentially no options to specifically name your samples or group reads across lanes. Read more about the underlying logic and options <a href="https://www.nextflow.io/docs/latest/reference/channel.html#fromfilepairs">here</a>.</p>
<p>Here are a few examples for common file naming patterns:</p>
<div class="tabbed-set tabbed-alternate" data-tabs="2:3"><input checked="checked" id="__tabbed_2_1" name="__tabbed_2" type="radio" /><input id="__tabbed_2_2" name="__tabbed_2" type="radio" /><input id="__tabbed_2_3" name="__tabbed_2" type="radio" /><div class="tabbed-labels"><label for="__tabbed_2_1">Illumina</label><label for="__tabbed_2_2">SRA/ENA</label><label for="__tabbed_2_3">Single-end data</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<p>Illumina paired-end file naming follows the convention:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-10-1"><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a>SampleName_SX_LYYY_R1_001.fastq.gz
</span><span id="__span-10-2"><a id="__codelineno-10-2" name="__codelineno-10-2" href="#__codelineno-10-2"></a>SampleName_SX_LYYY_R2_001.fastq.gz
</span></code></pre></div>
<p>Where SX is the sample number and LYYY the lane number.
These names can be parsed with the following pattern:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-11-1"><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a>&#39;/path/to/reads/*_R{1,2}_001.fastq.gz&#39;
</span></code></pre></div>
<p>resulting in sample names being parsed as <code>SampleName_SX_LYYY</code>. However, as mentioned, there is no way to parse common sample names across lanes so you would need to merge multi-lane reads before using this input approach.</p>
</div>
<div class="tabbed-block">
<p>Data downloaded from online archives like SRA or ENA follow the convention:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-12-1"><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a>SampleName_1.fastq.gz
</span><span id="__span-12-2"><a id="__codelineno-12-2" name="__codelineno-12-2" href="#__codelineno-12-2"></a>SampleName_2.fastq.gz
</span></code></pre></div>
<p>Which can be parsed with:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-13-1"><a id="__codelineno-13-1" name="__codelineno-13-1" href="#__codelineno-13-1"></a>&#39;/path/to/reads/*_{1,2}.fastq.gz&#39;
</span></code></pre></div>
</div>
<div class="tabbed-block">
<p>Collecting single-end data is also possible and fairly straightforward:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-14-1"><a id="__codelineno-14-1" name="__codelineno-14-1" href="#__codelineno-14-1"></a>&#39;/path/to/reads/*.fastq.gz&#39;
</span></code></pre></div>
</div>
</div>
</div>
</dd>
<dt><code>--reference_base</code> [default = null ]</dt>
<dd>
<p>The location where the pipeline references are installed on your system. This will typically be pre-set in your site-specific config file and is only needed when you run without one.</p>
<p>See our <a href="../installation/">installation guide</a> to learn how to install the references permanently on your system.</p>
</dd>
<dt><code>--outdir results</code> [default = results]</dt>
<dd>
<p>The location where the results are stored. Usually this will be <code>results</code> in the location from where you run the nextflow process. However, this option also accepts any other path in your file system(s).</p>
</dd>
<dt><code>--run_name Fubar</code> [default = null]</dt>
<dd>
<p>A mandatory name for this run, to be included with the result files.</p>
</dd>
</dl>
<h3 id="pcr-primers">PCR primers</h3>
<dl>
<dt><code>--list_primers</code> [ default = false]</dt>
<dd>
<p>Get a list of pre-configured primer sets.</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-15-1"><a id="__codelineno-15-1" name="__codelineno-15-1" href="#__codelineno-15-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>bio-raum/FooDMe2<span class="w"> </span>--list_primers
</span></code></pre></div>
</dd>
<dt><code>--primer_set</code> [default = null]</dt>
<dd>
<p>The name of the pre-configured primer set to use for read clipping. More sets will be added over time</p>
<p>Available options <a href="../../methods/methods/">here</a>.</p>
<p>A list of pre-configured primer sets is also available from the pipeline directly, with <code>--list_primers</code></p>
</dd>
<dt><code>--primers_fa</code> [default = null]</dt>
<dd>
<p>If you do not wish to use a pre-configured primer set, you can alternatively provide primer sequences in FASTA format. This option requires <code>--db</code> or <code>--blast_db</code> to choose the appropriate database to compare your data against.
Please make sure that your primer sequences only contain <a href="https://www.bioinformatics.org/sms/iupac.html">IUPAC-compliant</a> bases.</p>
</dd>
</dl>
<h3 id="database">Database</h3>
<p>Databases for taxonomic assignment can be specified in one of two ways - from the pre-installed references or as a user-supplied option.</p>
<dl>
<dt><code>--list_dbs</code></dt>
<dd>
<p>You can get a list of available databases and their origin as follows:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-16-1"><a id="__codelineno-16-1" name="__codelineno-16-1" href="#__codelineno-16-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>bio-raum/FooDMe2<span class="w"> </span>--list_dbs
</span></code></pre></div>
</dd>
<dt><code>--db</code> [default = null]</dt>
<dd>
<p>Use a pre-installed database (recommended!). Available options are (common choices in bold):</p>
<table>
<thead>
<tr>
<th>name</th>
<th>source</th>
</tr>
</thead>
<tbody>
<tr>
<td>a6</td>
<td>Midori</td>
</tr>
<tr>
<td>a8</td>
<td>Midori</td>
</tr>
<tr>
<td><strong>srna</strong></td>
<td>Midori</td>
</tr>
<tr>
<td><strong>lrna</strong></td>
<td>Midori</td>
</tr>
<tr>
<td><strong>co1</strong></td>
<td>Midori</td>
</tr>
<tr>
<td>co2</td>
<td>Midori</td>
</tr>
<tr>
<td>co3</td>
<td>Midori</td>
</tr>
<tr>
<td><strong>cytb</strong></td>
<td>Midori</td>
</tr>
<tr>
<td><strong>genbank</strong></td>
<td>NCBI</td>
</tr>
<tr>
<td>ncbi_its</td>
<td>NCBI</td>
</tr>
<tr>
<td><strong>its</strong></td>
<td>Unite</td>
</tr>
<tr>
<td>nd1</td>
<td>Midori</td>
</tr>
<tr>
<td>nd2</td>
<td>Midori</td>
</tr>
<tr>
<td>nd3</td>
<td>Midori</td>
</tr>
<tr>
<td>nd4l</td>
<td>Midori</td>
</tr>
<tr>
<td>nd5</td>
<td>Midori</td>
</tr>
<tr>
<td>nd6</td>
<td>Midori</td>
</tr>
<tr>
<td><strong>refseq</strong></td>
<td>RefSeq</td>
</tr>
</tbody>
</table>
<p>The underlying databases are obtained from <a href="https://www.reference-midori.info/">Midori</a>, <a href="https://unite.ut.ee/index.php">Unite</a> and <a href="https://ftp.ncbi.nlm.nih.gov/blast/db">NCBI</a>.</p>
</dd>
<dt><code>--blast_db</code> [ default = null]</dt>
<dd>
<p>Provide your own blast database. This requires that the database has valid taxonomy IDs included and should only be attempted by experienced users. Databases must be created with the options <code>--parse_seqids</code> and <code>--taxid_map</code> using the NCBI taxonomy.</p>
</dd>
<dt><code>--taxid_filter</code> [ default = null ]</dt>
<dd>
<p>In case you do not use a pre-configured <a href="#pcr-primers">primer_set</a> or wish to use it against a different taxonomic group, you will have to tell the pipeline a taxonomic group you wish to screen. The argument must be an ID from the <a href="https://www.ncbi.nlm.nih.gov/taxonomy">NCBI taxonomy</a>. Some common examples are:</p>
<table>
<thead>
<tr>
<th>Taxonomic group</th>
<th>NCBI ID</th>
</tr>
</thead>
<tbody>
<tr>
<td>Amniotes</td>
<td>32524</td>
</tr>
<tr>
<td>Mammals</td>
<td>40674</td>
</tr>
<tr>
<td>Insects</td>
<td>50557</td>
</tr>
<tr>
<td>Teleost fishes</td>
<td>32443</td>
</tr>
</tbody>
</table>
<p>Please note that the higher the node (i.e. the broader the search space), the more RAM will be required. This is not a concern for the single gene databases (e.g. Midori), but will be a significant factor when screening against GenBank NT. If you need to use GenBank NT and find that your jobs crash due to an out-of-memory error, consider using a shallower taxonomic node. </p>
</dd>
</dl>
<h2 id="advanced-options">Advanced options</h2>
<h3 id="benchmarking">Benchmarking</h3>
<p>It is possible to benchmark the pipelines performance against a set of known samples (e.g. for validation).
Therefore, predicted and expected components will be matched in a 'least distance' manner. A match will be deemed positive if the last common ancestor of both components
is at a maximum given rank and it's predicted (and expected) proportion in the sample is at least at a certain threshold.</p>
<p>Benchmarking is activated by providing following arguments:</p>
<dl>
<dt><code>--ground_truth</code> [default = false]</dt>
<dd>
<p>Path to a tab-delimited text file giving expected sample composition. The file must be formatted as follows:</p>
<div class="language-text highlight"><pre><span></span><code><span id="__span-17-1"><a id="__codelineno-17-1" name="__codelineno-17-1" href="#__codelineno-17-1"></a>sample  taxid   proportion
</span><span id="__span-17-2"><a id="__codelineno-17-2" name="__codelineno-17-2" href="#__codelineno-17-2"></a>S100    9303    0.9
</span><span id="__span-17-3"><a id="__codelineno-17-3" name="__codelineno-17-3" href="#__codelineno-17-3"></a>S100    9013    0.1
</span><span id="__span-17-4"><a id="__codelineno-17-4" name="__codelineno-17-4" href="#__codelineno-17-4"></a>S200    9303    0.8
</span><span id="__span-17-5"><a id="__codelineno-17-5" name="__codelineno-17-5" href="#__codelineno-17-5"></a>S200    9313    0.2
</span></code></pre></div>
<ul>
<li><code>sample</code>: Sample name</li>
<li><code>taxid</code>:Taxonomic identifier </li>
<li><code>proportion</code>: Expected proportion in the [0-1] interval, <strong>not</strong> in percent</li>
</ul>
</dd>
<dt><code>--benchmark_rank</code> [default = 'genus']</dt>
<dd>
<p>Maximum rank for a predicted match to be positive</p>
</dd>
<dt><code>--benchmark_cutoff</code> [default = 0.001]</dt>
<dd>
<p>Minimum proportion to be considered predicted</p>
</dd>
</dl>
<h3 id="pcr-primer-trimming">PCR primer trimming</h3>
<p>Some possible usage examples:</p>
<div class="tabbed-set tabbed-alternate" data-tabs="3:3"><input checked="checked" id="__tabbed_3_1" name="__tabbed_3" type="radio" /><input id="__tabbed_3_2" name="__tabbed_3" type="radio" /><input id="__tabbed_3_3" name="__tabbed_3" type="radio" /><div class="tabbed-labels"><label for="__tabbed_3_1">Default</label><label for="__tabbed_3_2">Custom primers</label><label for="__tabbed_3_3">Trim 3' ends</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<div class="language-bash highlight"><pre><span></span><code><span id="__span-18-1"><a id="__codelineno-18-1" name="__codelineno-18-1" href="#__codelineno-18-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>bio-raum/FooDMe2<span class="w"> </span>-profile<span class="w"> </span>standard,conda<span class="w"> </span>--input<span class="w"> </span>samples.csv<span class="w"> </span><span class="se">\\</span>
</span><span id="__span-18-2"><a id="__codelineno-18-2" name="__codelineno-18-2" href="#__codelineno-18-2"></a>--primer_set<span class="w"> </span>amniotes_dobrovolny<span class="w"> </span><span class="se">\\</span>
</span><span id="__span-18-3"><a id="__codelineno-18-3" name="__codelineno-18-3" href="#__codelineno-18-3"></a>--run_name<span class="w"> </span>cutadapt-test
</span></code></pre></div>
<p>This example uses a built-in primer set to perform primer removal.</p>
</div>
<div class="tabbed-block">
<div class="language-bash highlight"><pre><span></span><code><span id="__span-19-1"><a id="__codelineno-19-1" name="__codelineno-19-1" href="#__codelineno-19-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>bio-raum/FooDMe2<span class="w"> </span>-profile<span class="w"> </span>standard,conda<span class="w"> </span>--input<span class="w"> </span>samples.csv<span class="w"> </span><span class="se">\\</span>
</span><span id="__span-19-2"><a id="__codelineno-19-2" name="__codelineno-19-2" href="#__codelineno-19-2"></a>--primers_fa<span class="w"> </span>my_primers.fasta<span class="w"> </span><span class="se">\\</span>
</span><span id="__span-19-3"><a id="__codelineno-19-3" name="__codelineno-19-3" href="#__codelineno-19-3"></a>--db<span class="w"> </span>srna<span class="w"> </span><span class="se">\\</span>
</span><span id="__span-19-4"><a id="__codelineno-19-4" name="__codelineno-19-4" href="#__codelineno-19-4"></a>--run_name<span class="w"> </span>cutadapt-test
</span></code></pre></div>
<p>This example uses your custom primers, performs PCR primer site removal with cutadapt and performs taxonomic profiling against the srRNA database.</p>
</div>
<div class="tabbed-block">
<p><div class="language-bash highlight"><pre><span></span><code><span id="__span-20-1"><a id="__codelineno-20-1" name="__codelineno-20-1" href="#__codelineno-20-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>bio-raum/FooDMe2<span class="w"> </span>-profile<span class="w"> </span>standard,conda<span class="w"> </span>--input<span class="w"> </span>samples.csv<span class="w"> </span><span class="se">\\</span>
</span><span id="__span-20-2"><a id="__codelineno-20-2" name="__codelineno-20-2" href="#__codelineno-20-2"></a>--primer_set<span class="w"> </span>amniotes_dobrovolny<span class="w"> </span><span class="se">\\</span>
</span><span id="__span-20-3"><a id="__codelineno-20-3" name="__codelineno-20-3" href="#__codelineno-20-3"></a>--cutadapt_trim_3p<span class="w"> </span><span class="se">\\</span>
</span><span id="__span-20-4"><a id="__codelineno-20-4" name="__codelineno-20-4" href="#__codelineno-20-4"></a>--run_name<span class="w"> </span>cutadapt-test
</span></code></pre></div>
This example will additionally reverse complement your primer sequences and check for primer binding sites at both ends of each read.</p>
</div>
</div>
</div>
<dl>
<dt><code>--cutadapt_trim_3p</code> [ default = false ]</dt>
<dd>
<p>Use this option if you know that your read length is as long or longer than your PCR product. In this case, the reads will carry both the forward and reverse primer site - something that Cutadapt will normally fail to detect.</p>
</dd>
<dt><code>--cutadapt_trim_flex</code> [ default = false ]</dt>
<dd>
<p>Similar to <code>--cutadapt_trim_3p</code> but performs <strong>optional</strong> trimming on the 3' end. Usefull if you expect a mix of amplicons longer and shorter than the read length.</p>
</dd>
<dt><code>--store_reads</code> [ default = false ]</dt>
<dd>
<p>Emit the primer-trimmed reads into the result folder. This option is mostly useful to debug errors that are related to failed primer site removal. This option is set to false by default to save storage space.</p>
</dd>
<dt><code>--cutadapt_options</code> [ default = "" ]</dt>
<dd>
<p>Any additional options you feel should be passed to <a href="https://cutadapt.readthedocs.io/en/stable/reference.html#command-line-options">Cutadapt</a>. Use at your own risk.</p>
</dd>
</dl>
<h3 id="read-trimming">Read trimming</h3>
<dl>
<dt><code>--fastp_options</code> [ default = "-l 50 -3 --cut_tail_window_size 4 --cut_tail_mean_quatlity 25"]</dt>
<dd>
<p>Options for basic read trimming using <a href="https://github.com/OpenGene/fastp?tab=readme-ov-file#input-and-output">fastP</a>. You can overwrite the defaults if your data requires
it. Make sure to re-set any options you want to keep. </p>
</dd>
</dl>
<h3 id="clustering-configuration">Clustering configuration</h3>
<dl>
<dt><code>--amplicon_min_length</code> [ default = 70 ]</dt>
<dd>
<p>The minimum size an amplicon is expected to have. Data that falls below this threshold will be discarded. This option does not need to be touched for pre-configured primer profiles.</p>
</dd>
<dt><code>--amplicon_max_length</code> [ default = 100 ]</dt>
<dd>
<p>The maximum size an amplicon is expected to have. Data that lies above this threshold will be discarded. This option does not need to be touched for pre-configured primer profiles.</p>
</dd>
<dt><code>--non_overlapping</code> [ default = false]</dt>
<dd>
<p>Toggle read concatenation instead of merging with an overlapping sequence. A short N spacer will be added between the forward and reverse read sequences. Works for both VSEARCH and DADA2 with paired-end reads. This is useful in case long amplicons and/or short sequencing reads lead to R1 and R2 having no overlap. <strong>This applies to all reads.</strong> Trying to analyze datasets with a mixture of overlapping and non-overlapping reads is highly discouraged.</p>
<div class="admonition warning">
<p class="admonition-title">Non-overlapping reads and BLAST</p>
<p>Concatenation of non-overlapping reads with a N spacer still allows BLAST to find most similar sequences based on local alignment.
You should however keep in mind that the query coverage will then be drastically lower than for overlapping reads, since the non-sequenced portion of the amplicon will not be part of the alignement.
Make sure you choose a value for <code>--blast_qcov</code> that is adapted to your application.</p>
</div>
</dd>
<dt><code>--vsearch</code> [ default = false ]</dt>
<dd>
<p>The default tool to compute OTUs/ASVs is DADA2. Use this option to run VSEARCH instead (short reads only).</p>
</dd>
<dt><code>--vsearch_min_cov</code> [ default = 5 ]</dt>
<dd>
<p>The minimum amount of coverage required for an OTU to be created from the read data.</p>
</dd>
<dt><code>--vsearch_cluster_id</code> [ default = 98 ]</dt>
<dd>
<p>The percentage similarity for ASUs to be collapsed into OTUs. If you set this to 100, ASUs will not be collapsed at all, which will generate a higher resolution call set at the cost of added noise. In turn, setting this value too low may collapse separate species into "hybrid" OTUs.
The default of 98 seems to work quite well for our data, but will occasionally fragment individual taxa into multiple OTUs if sequencing error rate is high. For the TSV output, OTUs with identical taxonimic assignments will be counted as one, whereas the JSON output leaves this step to the user.</p>
</dd>
<dt><code>--remove_chimera</code> [ default = true ]</dt>
<dd>
<p>Toggle chimera filtering step on or off.</p>
</dd>
<dt><code>--max_expected_errors</code> [ default = 2 ]</dt>
<dd>
<p>Maximum allowed amount of expected errors in reads.</p>
</dd>
<dt><code>--max_ns</code> [ default = 0 ]</dt>
<dd>
<p>Maximum allowed number of undetermined ('N') nucleotides in reads.</p>
</dd>
<dt><code>--merging_max_mismatch</code> [ default = 1 ]</dt>
<dd>
<p>Maximum allowed number of mismatches in the overlap region for read merging. Only for paired-end reads.</p>
</dd>
</dl>
<h3 id="blast-configuration">BLAST configuration</h3>
<dl>
<dt><code>--blocklist</code></dt>
<dd>
<p>Provide a list of NCBI taxonomy IDs (one per line) that should be masked from the BLAST database (and thus the result). FooDMe2 uses a built-in <a href="https://raw.githubusercontent.com/bio-raum/FooDMe2/main/assets/blocklist.txt">block list</a> - but you can use this option to overwrite it, if need be. A typical use case would be a list of taxa that you know for a fact to be false positive hits. Consider merging your list with the built-in block list to make sure you mask previously identified problematic taxa. </p>
</dd>
<dt><code>--disable_low_complexity</code> [ default = false]</dt>
<dd>
<p>By default, Blast with filter/main low complexity sequences. If your amplicons have very low complexity, you may wish to set this option to disable the masking of low complexity motifs. Effectively deactivates DUST fileter and soft masking.</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-21-1"><a id="__codelineno-21-1" name="__codelineno-21-1" href="#__codelineno-21-1"></a>nextflow<span class="w"> </span>run<span class="w"> </span>bio-ram/FooDMe2<span class="w"> </span>
</span><span id="__span-21-2"><a id="__codelineno-21-2" name="__codelineno-21-2" href="#__codelineno-21-2"></a>-profile<span class="w"> </span>apptainer<span class="w"> </span><span class="se">\</span>
</span><span id="__span-21-3"><a id="__codelineno-21-3" name="__codelineno-21-3" href="#__codelineno-21-3"></a>--input<span class="w"> </span>samples.tsv<span class="w"> </span><span class="se">\</span>
</span><span id="__span-21-4"><a id="__codelineno-21-4" name="__codelineno-21-4" href="#__codelineno-21-4"></a>--disable_low_complexity<span class="w"> </span>...
</span></code></pre></div>
</dd>
<dt><code>--blast_evalue</code> [ default = "1e-20" ]</dt>
<dd>
<p>Maximal e-value of BLAST results</p>
</dd>
<dt><code>--blast_qcov</code> [ default = "100" ]</dt>
<dd>
<p>Amount of the query that has to be covered by a BLAST hit, in percent.</p>
</dd>
<dt><code>--blast_perc_id</code> [ default = "97" ]</dt>
<dd>
<p>Minimal identity level between query and BLAST hit, in percent</p>
</dd>
<dt><code>--blast_bitscore_diff</code> [ default = 4 ]</dt>
<dd>
<p>Maximal difference between the best BLAST hit's bitscore and the other hits to be kept.</p>
</dd>
<dt><code>--blast_min_consensus</code> [ default = 0.51 ]</dt>
<dd>
<p>Minimal consensus level between all BLAST results for a given query to be assigned to a taxonomic node.</p>
</dd>
</dl>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyrigth &copy; GNU-GPL - 2024 Grégoire Denay and Marc Höppner
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.instant", "navigation.tracking", "navigation.tabs", "navigation.tabs.sticky", "navigation.top", "navigation.path", "toc.follow", "content.code.copy", "content.code.annotate"], "search": "../../assets/javascripts/workers/search.f8cc74c7.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"alias": true, "default": "latest", "provider": "mike"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.c8b220af.min.js"></script>
      
    
  </body>
</html>