<!DOCTYPE html>
<html>
<head>
<meta charset="utf-8" />   <!--It is necessary to use the UTF-8 encoding with plotly graphics to get e.g. negative signs to render correctly -->
<meta name="viewport" content="width=device-width, initial-scale=1.0" />
<style>
    body {background-color: lightslategray; font-family: sans-serif; font-size: 10pt; }
    .header {  display: block; font-size: 10px; color: grey ; background-color: lightgrey; padding: 5px; }
    a {color: slategray; text-decoration: none;}
    a:hover {color: darkgrey ; text-decoration: none;}
    h1 strong { display: block; font-size: 75%; opacity: 0.25; }
    h2 {background: lightslategray; color: white; padding-left: 5px ;}
    .ground {background: white ; display: block; padding: 5px; height: fit-content}
    .impressum {background: lightgrey; display: block; padding: 5px; vertical-align: bottom; font-size: 10px;}
    .info { font-size: 10px; color: grey; }
    td, td, th { padding: 5px;}
    tr td[scope="row"] {background-color: lightgray;}
    tr td[scope="sample-id"] {background-color: lightgray; font-weight: bold;}
    tr td[scope="warn"] {background-color: rgb(240, 184, 80);}
    tr td[scope="pass"] {background-color: rgb(65, 156, 65); color: whitesmoke;} 
    tr td[scope="fail"] {background-color: rgb(236, 74, 74);}
    tr td[scope="missing"] {background-color: lightgreen;}
    tr th[scope="col"] {background-color: #505050; color: #fff;}
    tr th[scope="subcol"] {background-color: #888686; color: #fff;}
    .nav td { border-top: 1px solid; border-bottom: 1px solid ;}
    .versions { font-size: 9px; }
    .versions td { padding: 0px; }
    /* Tooltip container */
    .tooltip { position: relative; display: inline-block; border-bottom: 1px dotted black; }
    .tooltip .tooltiptext { visibility: hidden; width: 250px; background-color: darkgrey; color: black; text-align: center; padding: 5px 0; border-radius: 6px;  position: absolute;  z-index: 1; }
    .tooltip:hover .tooltiptext { visibility: visible; }
</style>
</head>

<body>

<div class="header">
    <a href="https://github.com/bio-raum">github.com/bio-raum</a> - bioinformatic resources for surveillance in food safety and public health. 
</div>

<div class="ground">

<h1>
    <div id="summary"></div>FooDMe2
    <strong>Taxonomic profiling of metabarcoding data</strong>
</h1>

<div id="navigation"></div>
<table  class="nav">
    <tr>
        <td><a href="#summary">Summary</a></td>
        <td><a href="#taxa">Taxonomic composition</a></td>
        <td><a href="#insertsizes">Insert sizes</a></td>
    </tr>
</table>
<!--
The summary table with all key metrics per sample
-->
<h2>Summary</h2>
<table>
    <tr>
        <th scope="col">Sample</th>
        <th scope="col"><div class="tooltip">Status<span class="tooltiptext">The overall analysis status: pass: ok to use, warn: potential issues found, fail: most probably not usable</span></div></th>
        <th scope="col"><div class="tooltip">Reads total<span class="tooltiptext">The number of reads before any processing</span></div></th>
        <th scope="col"><div class="tooltip">Reads Q30 (%)<span class="tooltiptext">Fraction of input reads >= Q30</span></div></th>
        <th scope="col"><div class="tooltip">Reads passing<span class="tooltiptext">The number of reads passing the adapter and quality trimming</span></div></th>
        <th scope="col"><div class="tooltip">Reads filtered<span class="tooltiptext">The number of reads not passing the adapter and quality trimming</span></div></th>
        <th scope="col"><div class="tooltip">Reads after clustering<span class="tooltiptext">The number of reads remaining after clustering</span></div></th>
        <th scope="col"><div class="tooltip">Chimeric reads<span class="tooltiptext">Reads classified as chimeria during the clustering</span></div></th>
        <th scope="col"><div class="tooltip">Insert size (bp)<span class="tooltiptext">The peak insert size</span></div></th>
    </tr>

<!--
Our data structure is a list of dictionaries; where each dictionary holds
all the information for one sample    
-->
{% for row in summary %}
    <tr>
        <td scope="sample-id">{{row.sample}}</td>
        <td scope={{row.status}}>{{row.status}}</td>
        <td>{{row.reads_total}}</td>
        <td>{{row.reads_q30}}</td>
        <td>{{row.reads_passing}}</td>
        <td>{{row.reads_filtered}}</td>
        <td>{{row.reads_after_clustering}}</td>
        <td>{{row.reads_chimera}}</td>
        <td>{{row.insert_size}}</td>
    </tr>
{% endfor %}
</table>

<p></p>
<a href="#navigation">top</a>

<div id="taxa"></div>
<h2>Taxonomic composition</h2>

{{Taxa}}

<div class="info">
*Showing taxa with an abundance of >= 1%. Please consult the detailed reports in JSON and XLSX format for a finer resolution. 
</div>

<p></p>
<a href="#navigation">top</a>

<div id="insertsizes"></div>
<h2>Insert size distribution</h2>

{{Insertsizes}}

<p></p>
<a href="#navigation">top</a>

<div id="software"></div>
<h2>Software versions</h2>

</div>

<div class="impressum">
    Report generated by bio-raum/FooDMe2. Please check out our <a href="https://bio-raum.github.io/FooDMe2/">documentation</a>. 
</div>
</body>
</html>